radix-engine/src/transaction/transaction_executor.rs

```diff
@@ -782,32 +782,23 @@ where
     }
 
     fn determine_result_type(
-        mut interpretation_result: Result<Vec<InstructionOutput>, TransactionExecutionError>,
+        interpretation_result: Result<Vec<InstructionOutput>, TransactionExecutionError>,
         fee_reserve: &mut SystemLoanFeeReserve,
     ) -> TransactionResultType {
         // A `SuccessButFeeLoanNotRepaid` error is issued if a transaction finishes before
         // the SYSTEM_LOAN_AMOUNT is reached (which trigger a repay event) and even though
         // enough fee has been locked.
         //
         // Do another `repay` try during finalization to remedy it.
-        if let Err(err) = fee_reserve.repay_all() {
-            if interpretation_result.is_ok() {
-                interpretation_result = Err(TransactionExecutionError::RuntimeError(
-                    RuntimeError::SystemModuleError(SystemModuleError::CostingError(
-                        CostingError::FeeReserveError(err),
-                    )),
-                ));
-            }
-        }
+        let final_repay_result = fee_reserve.repay_all();
 
         match interpretation_result {
-            Ok(output) => {
-                if fee_reserve.fully_repaid() {
-                    TransactionResultType::Commit(Ok(output))
-                } else {
-                    panic!("Manifest interpretation result was okay, but fee reserve wasn't fully repaid.")
+            Ok(output) => match final_repay_result {
+                Ok(_) => TransactionResultType::Commit(Ok(output)), // success and system loan repaid fully
+                Err(_) => {
+                    TransactionResultType::Reject(RejectionReason::SuccessButFeeLoanNotRepaid)
```

Comment:
I wonder if we should make it `RejectionReason::SuccessButFeeLoanNotRepaid(FeeReserveError)`, and map out the `err` here?

I think on balance we're fine as is, but interested in your take :).